ipq40xx: mtd: mac80211: Add Support for Linksys EA8300 (Dallas)

jeffsf opened the pull request:
As this PR touches code that should be transparent for other devices, but does change generic or other platform code, it has been split into three commits for future clarity.

Bring-up is on Linux 4.14 and includes the known limitations of the platform, in particular the "hard-wiring" of the Ethernet interfaces and the switch ports.

Device is named "ea8300" down through the DTS files. There has been talk that it should be named "dallas" from the DTS up to some level. I am willing to change that at the lower levels if the project decides to adopt that convention. (I prefer that the end-user see "ea8300" as I believe that is what a non-technical user would expect.)

ath10k board-file components for the three radios and four regions (AH, EU, IC, FCC) are ready for upstream submission. I have not submitted them due to the possible renaming of the device as just mentioned. 

Changes to `mtd resetbc` have additionally been checked by a third party on mvebu/mamba. 

The device will function without _mac80211: Select Better Ch/VHT for Sub-Band Radios_ (presently https://github.com/openwrt/openwrt/commit/b219e82a260536b44b661139ad461fee533a64f3), though the default `/etc/config/wireless` will set channel 36, which is out of the operational band allowed by one of the radio's pre-cal and board data.

The issue noted with respect to `firstboot` not resetting when used under failsafe is tracked at
* https://bugs.openwrt.org/index.php?do=details&task_id=468
and discussed at at least 
* https://forum.openwrt.org/t/solved-in-recovery-mode-firstboot-fails-jffs2reset-writing-dev-ubi0-1-failed-operation-not-permitted/28097

_Note that it is not solved, but that a work-around to flash without preserving config met the OP's need._

This PR overtakes the following which have now been closed:
* https://patchwork.ozlabs.org/patch/1060461/
* https://patchwork.ozlabs.org/patch/1060597/

jeffsf committed:
mtd: base-files: Unify dual-firmware devices (Linksys)

Consistently handle boot-count reset and upgrade across
ipq40xx, ipq806x, kirkwood, mvebu

Dual-firmware devices often utilize a specific MTD partition
to record the number of times the boot loader has initiated boot.

Most of these devices are NAND, typically with a 2k erase size.
When this code was ported to the ipq40xx platform, the device in hand
used NOR for this partition, with a 16-byte "record" size. As the
implementation of `mtd resetbc` is by-platform, the hard-coded nature
of this change prevented proper operation of a NAND-based device.

* Unified the "NOR" variant with the rest of the Linksys variants

* Added logging to indicate success and failure

* Provided a meaningful return value for scripting

* "Protected" the use of `mtd resetbc` in start-up scripts so that
   failure does not end the boot sequence

* Moved Linksys-specific actions into common `/etc/init.d/bootcount`

For upgrade, these devices need to determine which partition to flash,
as well as set certain U-Boot envirnment variables to change the next
boot to the newly flashed version.

* Moved upgrade-related environment changes out of bootcount

* Combined multiple flashes of environment into single one

* Current-partition detection now handles absence of `boot_part`

Runtime-tested: Linksys EA8300

Signed-off-by: Jeff Kletsky <git-commits@allycomm.com>

jeffsf committed:
ipq40xx: Add support for Linksys EA8300 (Dallas)

The Linksys EA8300 is based on QCA4019 and QCA9888 and provides three,
independent radios. NAND provides two, alternate kernel/firmware
images with fail-over provided by the OEM U-Boot.

Installation:

  "Factory" images may be installed directly through the OEM GUI.

Hardware Highlights:

  * IPQ4019 at 717 MHz (4 CPUs)
  * 256 MB NAND (Winbond W29N02GV, 8-bit parallel)
  * 256 MB RAM
  * Three, fully-functional radios; `iw phy` reports (FCC/US, -CT):
      * 2.4 GHz radio at 30 dBm
      * 5 GHz radio on ch. 36-64 at 23 dBm
      * 5 GHz radio on ch. 100-144 at 23 dBm (DFS), 149-165 at 30 dBm
      #{ managed } <= 16, #{ AP, mesh point } <= 16, #{ IBSS } <= 1
      * All two-stream, MCS 0-9
  * 4x GigE LAN, 1x GigE Internet Ethernet jacks with port lights
  * USB3, single port on rear with LED
  * WPS and reset buttons
  * Four status lights on top
  * Serial pads internal (unpopulated)

  "Linksys Dallas WiFi AP router based on Qualcomm AP DK07.1-c1"

Implementation Notes:

  The OEM flash layout is preserved at this time with 3 MB kernel and
  ~69 MB UBIFS for each firmware version. The sysdiag (1 MB) and
  syscfg (56 MB) partitions are untouched, available as read-only.

Serial Connectivity:

  Serial connectivity is *not* required to flash.

  Serial may be accessed by opening the device and connecting
  a 3.3-V adapter using 115200, 8n1. U-Boot access is good,
  including the ability to load images over TFTP and
  either run or flash them.

  Looking at the top of the board, from the front of the unit,
  J3 can be found on the right edge of the board, near the rear

      |
   J3 |
  |-| |
  |O| | (3.3V seen, open-circuit)
  |O| | TXD
  |O| | RXD
  |O| |
  |O| | GND
  |-| |
      |

Unimplemented:

    * serial1 "ttyQHS0" (serial0 works as console)
    * Bluetooth; Qualcomm CSR8811 (potentially conected to serial1)

Other Notes:

    https://wikidevi.com/wiki/Linksys_EA8300 states

        FCC docs also cover the Linksys EA8250. According to the
	RF Test Report BT BR+EDR, "All models are identical except
	for the EA8300 supports 256QAM and the EA8250 disable 256QAM."

Signed-off-by: Jeff Kletsky <git-commits@allycomm.com>

ynezz commented:
> This PR overtakes the following unaddressed RFCs/patches which may be closed
> https://patchwork.ozlabs.org/patch/1060461/
> https://patchwork.ozlabs.org/patch/1060597/

I've done as requested, but honestly I quite don't get this `unaddressed RFCs` note, since the patches were sent to the mailing list just a few days ago.

jeffsf commented:
Thanks -- only that they didn't need to be unwound from anywhere and could be dropped from further attention.

chunkeey commented on package/kernel/mac80211/files/lib/wifi/mac80211.sh:
```diff
@@ -81,13 +85,39 @@ detect_mac80211() {
-			iw phy "$dev" info | grep -q 'VHT Capabilities' && htmode="VHT80"
-		}
+
+			channel=$( echo "${iw_phy_info}" | \
```
There's a problem here (also with the old code - but nobody complained because of it).
You can't really have a channel at that time without having set the country (code) as well.
Problem is that the country setting affects the channel lists to a major degree and currently, it's set much later in drv_mac80211_setup().

What I think would make more sense is to tell hostapd to select a channel (for example by setting ```channel=acs_survey``` in the default configuration and improve the channel selection
there if necessary.

chunkeey commented on package/kernel/mac80211/files/lib/wifi/mac80211.sh:
```diff
@@ -81,13 +85,39 @@ detect_mac80211() {
+
+			if echo "${iw_phy_info}" | fgrep -q 'VHT Capabilities'
+			then
+				case "$channel" in
```
From afar this looks like you just want to fix something for your specific device in just the US?
But is this a problem? From what I know people just do set a channel when they are setting up their
wireless security key.

jeffsf replied:
The impetus here is primarily that the config generated isn't "operational" for the radio that is limited to ch.100-165. In LuCI, you can press "Enable" and nothing happens. I agree that it is likely that the user will select a channel (though far from certain, given the number of "Why doesn't 'auto' do anything automatically?" posts seen on the forum).

As "5180" appears as a string in the output of `iw phy phy0 info` with `ieee80211-freq-limit = <5490000 5835000>;` applied, the `master` implementation should at least pick up that it should be configured as a 5 GHz radio. 

I have confirmed that the `master` version identifies and provides config for all three radios, even with the freq. limits in place.

I have confirmed that LuCI shows the limited set of channels in is drop-downs.

Things get even more interesting with the -AH cal files and JP as a country code, where it looks like there aren't any available channels for the second radio (from poking at it, as I haven't found a tool to decode these pre-cal/board combinations, only the older all-in-ART ones).

I did try `channel=acs_survey` and it seemed to crash the wireless, at least when used in the context of `option channel 'acs_survey'` in `/etc/config/wireless` (assuming that the parameter value would be copied verbatim to the hostapd config). I did not pursue it further.

While not the best user experience, if there are no objections, at this time I'd like to put 802.11 config generation off into a separate exploration. The more I look at it, the more challenging it becomes to do in a way that is significantly better for the less-experienced user.

chunkeey replied:
> The impetus here is primarily that the config generated isn't "operational" for the radio that is limited to ch.100-165. In LuCI, you can press "Enable" and nothing happens. I agree that it is likely that the user will select a channel (though far from certain, given the number of "Why doesn't 'auto' do anything automatically?" posts seen on the forum).

This is all well and true. But from experience with DFS and regdomain, I think the solution here is to make "auto" work. And since your patch touches on this very area, I think a closer look at the userspace wpad (hostapd / wpa_supplicant) and the kernel (mt76) is duly warranted here. Because what else can you do that will work for the default?
You already touched on the regdb.txt issue with JP and AH. Furthermore channels can be dynamically disabled because of an operating radar nearby so that device has to either move to a different channel or cease operation completely. Really, I think the hostapd/wpad is the only place where a sensible channel and operation mode(HT20/40, VHT40/80/80+80/160) can be performed.

(Of course, if you don't want to deal with that side of issues as well, you can just do what other did and just drop the patch and it will be the end of that).

> As "5180" appears as a string in the output of iw phy phy0 info with ieee80211-freq-limit = <5490000 5835000>; applied, the master implementation should at least pick up that it should be configured as a 5 GHz radio.

I think this is something that @ecsv might be interested in as well. The A62's 5GHz wifis also have set limits for their operational range due to in-line band filters. So I wonder how the OpenMesh deals with the problem or if they are just providing a customized /etc/config/wireless in their factory image and the wifi detection is skipped.

> I did try channel=acs_survey and it seemed to crash the wireless, at least when used in the context of option channel 'acs_survey' in /etc/config/wireless (assuming that the parameter value would be copied verbatim to the hostapd config). I did not pursue it further.

Yes, ```channel=acs_survey``` is the hostapd.conf option. From what I know the /e(tc)/c(config)/wireless option is ```auto```. But I think it's also possible to omit the channel option in /e/c/wireless. 

> While not the best user experience, if there are no objections, at this time I'd like to put 802.11 config generation off into a separate exploration. The more I look at it, the more challenging it becomes to do in a way that is significantly better for the less-experienced user.

True, I was at odds with the wifi detection code in the past as well 5f8f8a366136a07df661e31decce2458357c167a 5e35b4562fb73aaa1c75a3404fed2efb6afe03f8 . 
From what I remember the issue with the channel selection also creeped up during the discussion, but there has been no push to do anything about it.

ecsv replied:
> I think this is something that @ecsv might be interested in as well. The A62's 5GHz wifis also have set limits for their operational range due to in-line band filters. So I wonder how the OpenMesh deals with the problem or if they are just providing a customized /etc/config/wireless in their factory image and the wifi detection is skipped.

There is no OpenMesh anymore - all part of Datto Networking now.

I am not involved in this part but they are using a replacement of netifd at the moment (and no Luci). Something which does additional checks before accepting channels for specific radios and so on. And the wireless configuration is also (partially) done using some different mechanism (don't ask me why).

chunkeey commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -18,7 +18,7 @@ endef
 # You can find instructions and examples on the linux-wireless wiki:
 # <https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles>
-ALLWIFIBOARDS:=alfa-network_ap120c-ac asus_map-ac2200 avm_fritzbox-7530 avm_fritzrepeater-3000 engenius_eap1300 engenius_ens620ext linksys_ea6350v3 qxwlan_e2600ac
+ALLWIFIBOARDS:=alfa-network_ap120c-ac asus_map-ac2200 avm_fritzbox-7530 avm_fritzrepeater-3000 engenius_eap1300 engenius_ens620ext linksys_ea6350v3 qxwlan_e2600ac linksys_ea8300
```
```
# Please send a mail with your device-specific board files upstream.
# You can find instructions and examples on the linux-wireless wiki:
# <https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles>
```

chunkeey commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -47,12 +47,16 @@ Don't install it for any other device!
 endef
 
-#$(eval $(call generate-ipq-wifi-package,<devicename>,<filename>,<displayname>))
+#$(eval $(call generate-ipq-wifi-package,<devicename>,<filename>,<displayname>[,addl-path, addl-file]))
```
hm, wouldn't it be better to extend the 

generate-ipq-wifi-package so it always takes the wifi-chip and board-2.bin path

and do customized helpers for the IPQ4019, QCA9888 (and QCA9984)

generate-ipq-wifi-package-ipq4019
generate-ipq-wifi-package-qca9888
(generate-ipq-wifi-package-qca9984 - no need yet, feel free to skip that)

Then you could have two packages for your EA8300. One for the QCA4019 board-2.bin 
and the other one for the QCA9888. The biggest hurdle would be to make it so that they
don't conflict with one another. But I think this is the cleanest way to do it, apart from
splitting the board-2.bin from the ath10k-firmware packages into separate packages.
What do you think?

jeffsf replied:
I think I left the biggest challenge for last 😉 

Not so much in a technical sense, but in a philosophy sense. 

I've got no concerns about needing to specify two (or even three) "packages" to get the proper board data into the build for a three-radio device. Only the device maintainers should ever have to worry about it, as they should be "default" for the device and just appear in the build.

I agree that there is benefit in making things modular and reusable. I haven't been able to convince myself that the ipq40xx platform is the only one that will be impacted. It wouldn't surprise me that as the performance demands on the radios increase, the need to specifically tune/tweak the parameters increases as well. In some sense, it could be "The new normal" that any new, mid- or upper-range device "needs" its custom data for at least as long as it takes for the files to make their way upstream and back again. The complexities of two radios on a single band just add to that feeling.

I'm still thinking what the advantage is, if any, of tying it to a platform, rather than having it available based on the use of a specific board and/or radio chip. I'm going to sleep on that one. 

Am I correct in assuming that the use of this approach is supposed to be only an interim thing until the properly "indexed" board-file segments are incorporated in the upstream sources? If so, then any "transition" package wouldn't be needed as newly ported devices could go with a new scheme, and the ones on an older scheme should naturally fall away as the segments are incorporated upstream.

chunkeey replied:
> I agree that there is benefit in making things modular and reusable. I haven't been able to convince myself that the ipq40xx platform is the only one that will be impacted. It wouldn't surprise me that as the performance demands on the radios increase, the need to specifically tune/tweak the parameters increases as well. In some sense, it could be "The new normal" that any new, mid- or upper-range device "needs" its custom data for at least as long as it takes for the files to make their way upstream and back again. The complexities of two radios on a single band just add to that feeling.

IPQ40XX is not the only platform, all wave v2 (and the older v1 customer products too) hardware is. You can extract the board-2.bin from Kvalo's ath10k-firmware repository and find tons of special boarddata for specific devices. As for OpenWrt. The main driving force behind the package was due to OpenMesh A42. However, somehow the boarddata files that @ecsv sent upstream were merged in a very timely fashion before this became a problem there. That said, Kalle does queue up the boarddata, you can find that he posted a mail of his latest queue on the mailing-list.
 
https://www.mail-archive.com/ath10k@lists.infradead.org/msg09966.html

> I'm still thinking what the advantage is, if any, of tying it to a platform, rather than having it available based on the use of a specific board and/or radio chip. I'm going to sleep on that one.

Yes, if you want you can make the ipq-wifi packages per platform. But you don't have to, personally I would wait around until one ath79 or ipq806x (upcoming ipq8074) user needs it. 

> Am I correct in assuming that the use of this approach is supposed to be only an interim thing until the properly "indexed" board-file segments are incorporated in the upstream sources? If so, then any "transition" package wouldn't be needed as newly ported devices could go with a new scheme, and the ones on an older scheme should naturally fall away as the segments are incorporated upstream.

Yes, we cleaned out the ipq-wifi package when the boarddata moved to upstream. See
25a72f5a01f80116d4cd0aafd99c0fd091be3fbd
e6bd568051c7ca77a59783fe50203f6e2a427f19
dceee8cc09a54f96dee82a6e95e3e77a6dd7ac13

chunkeey commented on target/linux/ipq40xx/base-files/etc/board.d/01_leds:
```diff
@@ -41,6 +41,9 @@ engenius,ens620ext)
 	ucidef_set_led_netdev "lan2" "LAN2" "${boardname}:green:lan2" "eth1"
 	;;
+linksys,ea8300)
+	ucidef_set_led_usbport "usb" "USB" "ea8300:green:usb" "usb1-port1"
```
I think you could convert this ucidef_ to device-tree method. See 
ef5ff08662532a6b1770214569e4c27b05e1dc35

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -93,6 +93,14 @@ ath10kcal_is_caldata_valid() {
 
 board=$(board_name)
 
+echo "$0: Extracting $FIRMWARE for $board"
```
Is this echo old remaining debugging code? Please remove it.

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -93,6 +93,14 @@ ath10kcal_is_caldata_valid() {
+echo "$0: Extracting $FIRMWARE for $board"
+
+case "$board" in
+	linksys,ea8300)
```
Please remove this case "$board"... esac (see below)

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -115,6 +123,11 @@ case "$FIRMWARE" in
+	linksys,ea8300)
+		ath10kcal_extract "0:ART" 36864 12064
+		# OEM assigns 4 sequential MACs
+		ath10kcal_patch_mac_crc $(macaddr_setbit_la $(macaddr_add "${base_mac}" 4))
```
Is it necessary to have macaddr_setbit_la here or is it because in the stock firmware "eth0" and that wifi shares the same MAC?

As for the ${base_mac}.  
I would just follow nbg6617,wre6606 there and put the ```$(cat /sys/class/net/eth0/address)```
into the ath10kcal_patch_mac_crc parameter.

```ath10kcal_patch_mac_crc $(macaddr_add $(cat /sys/class/net/eth0/address) _X)```

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -156,6 +169,10 @@ case "$FIRMWARE" in
 		;;
+	linksys,ea8300)
+		ath10kcal_extract "0:ART" 4096 12064
+		ath10kcal_patch_mac_crc $(macaddr_add "${base_mac}" 2)
```
Same as above.

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -209,6 +226,10 @@ case "$FIRMWARE" in
 		;;
+	linksys,ea8300)
+		ath10kcal_extract "0:ART" 20480 12064
+		ath10kcal_patch_mac_crc $(macaddr_add "${base_mac}" 3)
```
Same as above.

chunkeey commented on target/linux/ipq40xx/base-files/etc/init.d/bootcount:
```diff
@@ -10,5 +10,8 @@ start() {
 			echo -e "bootcount\nchanged\n" | /usr/sbin/fw_setenv -s -
 		;;
+	linksys,ea8300)
+		mtd resetbc s_env || true
```
Hm, target/linux/ipq40xx/base-files/etc/init.d/zlinksys_recovery also does a ```mtd resetbc s_env```.

@jeffsf @pepe2k @NoTengoBattery 
Wouldn't it be a good idea to merge target/linux/ipq40xx/base-files/etc/init.d/zlinksys_recovery and
target/linux/ipq40xx/base-files/etc/init.d/bootcount ?

pepe2k replied:
@chunkeey 

> Wouldn't it be a good idea to merge target/linux/ipq40xx/base-files/etc/init.d/zlinksys_recovery and
> target/linux/ipq40xx/base-files/etc/init.d/bootcount ?

ACK, all of that should be in `bootcount`.
Also, I would remove all echos and comments from the code.

NoTengoBattery replied:
If you want to merge the files... Please read the note in `zlinksys_recovery`. The file is called that way on purpose, as the purpose is to run the script at the very end of the boot process.

pepe2k replied:
Because...?

NoTengoBattery replied:
Do whatever you want.

pepe2k replied:
`bootcount` also executes with `START=99` so I don't understand your previous comment. Can't see anything special inside `zlinksys_recovery` which would prevent merging its code into `bootcount`.

chunkeey replied:
I think what @NoTengoBattery is refering to the comment in zlinksys_recovery there.

```
# Why? Because the custom sysupgrade script for the device will turn on
# auto_recovery to "yes". And it's the job of this script to set the
# boot boot_count to 0 and then disable auto_recovery, as that condition
# means that the stable release went well.
```

He wants to say that the recovery step has to be the last one because if some other S99* fails after the script has run, the device might still be a "dead duck". So only if the zlinksys_recovery is truly the last thing to run, the firmware recovery mechanism as a whole would work as intended.

jeffsf replied:
At least from what I have seen in the EA8300 and based on [some discussion on the EA6350v3 thread](https://forum.openwrt.org/t/add-support-for-linksys-ea6350-v3/18907/288?u=jeff
):

* The factory U-Boot appears to only use the boot counter partition to detect boot loops, not `boot_count` (I no longer set this on the EA8300, as it unnecessarily writes the U-Boot environment on every boot.)
* Disabling `auto_recovery` is probably a bad idea in any situation as the whole point is that if something is wrong with one firmware ("oops, I wrote over the kernel"), the other boots to allow more graceful recovery.

EA6350v3 (from https://openwrt.org/toh/linksys/linksys_ea6350_v3)
```
bootcmd=if test $boot_part = 1; then run bootpart1; else run bootpart2; fi
```
EA8300
```
bootcmd=if test $auto_recovery = no; then bootipq; elif test $boot_part = 1; then run bootpart1; else run bootpart2; fi
```
(note that `bootipq` did not seem functional on my EA8300; perhaps another bit of development cruft)

If there was a problem with leaving `auto_recovery = yes`, I would expect that one of the EA6350v3 users would have long ago mentioned it.

On the WIP responding to the very helpful comments on this PR, I have moved the setting of `auto_recovery` out of the every-boot process and into the sysupgrade process where, in my opinion, it belongs. This way the U-Boot environment partition is not written on every boot, saving flash wear. It is set when it is most likely to be needed, just before a potentially "bad" flash. As `boot_part` is being written at this time (and/or `bootcmd`, depending on board), there is no additional flash required (as all changes to the U-Boot environment are now done in a single call to `fw_setenv`).

chunkeey commented on target/linux/ipq40xx/base-files/lib/upgrade/linksys.sh:
```diff
@@ -1,26 +1,51 @@
-	echo "$target_firmware"
+	case $cur_boot_part in
+	1)
+		fw_setenv -s - <<EOF
```
From what I know you could use something like

```fw_setenv -s - <<- EOF```
to properly indent the following command list. I think it does look better, but I'm fine either way.

chunkeey commented on target/linux/ipq40xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -62,18 +62,19 @@ platform_do_upgrade() {
+	linksys,ea8300)
 		platform_do_upgrade_linksys "$ARGV"
 		;;
+	meraki,mr33)
```
Thank you for fixing the order!

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+		led-failsafe = &led_wps;
+		led-running = &led_linksys;
+		led-upgrade = &led_world;
+
```
Space?

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+		button@0 {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			linux,input-type = <0x1>;
```
EV_KEY ?

That said this is the default so you don't need to specify it.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+		button@1 {
+			label = "wps";
+			linux,code = <KEY_WPS_BUTTON>;
+			linux,input-type = <0x1>;
```
Same as above. You can drop linux,input-type

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+	//
+
+	chosen {
+		bootargs-append = " root=/dev/ubiblock0_0 rootfstype=squashfs ro clk_ignore_unused";
```
please drop "clk_ignore_unused". See the "admin-guide/kernel-parameters.txt"

"This parameter is useful for debug and development, but should not be needed
on a platform with proper driver support.  For more information, see Documentation/driver-api/clk.rst."

("ro" is kind of a odd but given the bootargs it's ok to leave it there)

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+
+		serial0 = &blsp1_uart1;
+	};
+
```
extra space? is this a recommendation or just your style?

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+	};
+
+
+	gpio_keys {
```
Please choose a generic node name. As per entry 5 on the https://openwrt.org/submitting-patches#dts_checklist Otherwise @ynezz will be mad.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+
+	soc {
+
+		serial@78af000 {		// blsp1_uart1
```
This node has existing labels? You could save the indents by moving this out of the main/soc nodes.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+		};
+
+
+		qpic-nand@79b0000 {		// nand
```
This node has existing labels. You could save the indents by moving this out of the main/soc nodes.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+					#size-cells = <1>;
+
+					partition@0 {
+						label = "0:SBL1";
```
Any reason to keep the "0:" prefix? From what I can tell, this is just qualcommins that they would use
if you have multiple flash chips. However even linksys decided to not care so we could at least have
a uniform label names (all lowercase and without prefix)

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+
+
+		i2c@78b7000 {			// blsp1_i2c3
+			status = "ok";
```
Please use "okay". As "ok" is deprecated.

However, I have a question: Is the i2c connected to anything at all?
If not please leave it disabled (by removing this i2c node altogether)

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+			pinctrl-names = "default";
+		};
+
+		spi@78b5000 {			// blsp1_spi1
```
Is something connected here? if not, please remove the node.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+			status = "okay";
+		};
+
+		spi@78b6000 {			// blsp1_spi2
```
Is something connected here? if not, please remove the node.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+			status = "okay";
+		};
+
+		rng@22000 {
```
Could you please order the nodes by the unit-address (22000) in ascending order?

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+				bias-disable;
+			};
+
+			nand_pins: nand-pins {
```
bias are missing. You could copy the node from the qcom-ipq4019-fritzbox-7530.dts.

jeffsf replied:
Using the bias settings from that device here is troublesome. 

While "only a little green light" that fails as a result makes it a red flag for me. The EA8300, the reference board variant it is based on, `qcom-ipq40xx-ap.dk07.1.dtsi`, and those similar have a different set of NAND pins[1] than the fritzbox-7530 suggested. gpio61, listed in the Fritzbox as a NAND pin, is used for LED control on the EA8300.

In searching through the Linux sources for 4.14 and 4.19 both, as well as the OEM GPL drop for the EA8300, I have not been able to find evidence of bias for the NAND pins being explicitly set in the DTS. Given that U-Boot was able to successfully read from the NAND to boot and write to it to change the environment and boot counter, I would guess that it is setting up the bias properly. Linksys did not release the U-Boot source so I do not have access to this.

I'd rather not put in information that I can't confirm (bias for NAND pins) when so much gets copied "blindly" from device to device. 

As I don't have access to an IPQ4019 data sheet, is there information available that I could use to examine one or more registers to confirm what the bias settings are from the running device?

[1] See, for example
http://downloads.linksys.com/downloads/gpl/EA8300_v1.1.3.184925_GPL.tar.gz
https://elixir.bootlin.com/linux/v4.19.36/source/arch/arm/boot/dts/qcom-ipq4019-ap.dk07.1.dtsi
https://elixir.bootlin.com/linux/v4.19.36/source/arch/arm/boot/dts/qcom-ipq4019-ap.dk04.1.dtsi

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:
```diff
@@ -0,0 +1,394 @@
+			qcom,usb-hsphy-mode-select = <TCSR_USB_HSPHY_HOST_MODE>;
+		};
+
+		ess_tcsr@1953000 {  // May not be needed for ipqess / qca8k
```
(It is needed there as well. But this is for another day)

chunkeey commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -230,6 +230,21 @@ define Device/linksys_ea6350v3
+	$(call Device/FitzImage)
+	DEVICE_TITLE := Linksys EA8300 (Dallas)
+	DEVICE_DTS := qcom-ipq4019-linksys_ea8300
+	KERNEL_SIZE := 3145728    # 0x5800000 - 0x5500000
```
```KERNEL_SIZE := 3072k``` or ```KERNEL_SIZE := 3m``` would work as well.

chunkeey commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -230,6 +230,21 @@ define Device/linksys_ea6350v3
+	DEVICE_TITLE := Linksys EA8300 (Dallas)
+	DEVICE_DTS := qcom-ipq4019-linksys_ea8300
+	KERNEL_SIZE := 3145728    # 0x5800000 - 0x5500000
+	IMAGE_SIZE := 89128960    # 0x5800000
```
85m ?

chunkeey reviewed (changes requested):
> IPQ4019 at 768 MHz (4 CPUs)

Is this in the EA8300 spec somewhere? Because the cpufreq is limited to just 710/716,666.. MHz

>  * Serial pads internal (unpopulated)

Can you please add the pin layout there? (what is GND,RX/TX,VCC?)
(And add a 3V3 warning so people don't accidentally connect RS232-level
devices)

>    * Current (4.14) ipq40xx switch driver "hard-wires"
>      eth0 to LAN1-4 ports and eth1 to WAN port

Well, the ar40xx allows for some limited changes. For example you could swap eth0/eth1
ports etc. But it's not like with the gmac / ar8216 on the ath79. 

> `firstboot` under failsafe does not erase UBI volume
>   (not specific to this device)

Maybe it doesn't work as you would expect it? Currently it works like this: firstboot (which is a wrapper for jffs2reset) just writes a "0xDEADC0DE" magic into the ubi rootfs_data volume. But it does not
erase the volume. That "marker" is then picked up during the overlay mount process (on normal boot) which would happen after a reboot for example, which causes it to erase the data in the rootfs_data before mounting the now "empty" volume. 

>    * ART data extracted "too late" on first boot
>      (not specific to this device)

Is it still happening, even with a69e101ed1169f562fc030a783cd997d3f066b16 ?

> Unimplemented:
>  * serial1 "ttyQHS0" (serial0 works as console)
Are there pads/unpopulated headers on the board for that? Or is the bluetooth:

>    * Bluetooth; Qualcomm CSR8811
Connected via this UART? Also do you have an idea what's missing for the CSR8811?

No 4.19 patches?

jeffsf committed:
firmware/ipq-wifi: Extend for multi-chip boards

This package provides board-specific reference ("cal") data
on an interim basis until included in the upstream distros

While originally conceived for IPQ4019-based boards, similar needs
are appearing with three-radio devices. For some of these devices,
both a board-2.bin file needs to be supplied both for the IPQ4019
as well as for the other radio on the board.

This patch allows new or multiple overrides to be specified by:

  * Adding board name to ALLWIFIBOARDS
  * Placing file(s) in this directory named as
      board-<devicename>.<qca4019|qca9888|qca9984>
  * Adding
      $(eval $(call generate-ipq-wifi-package,<device>,<display name>))

(along with suitable package selection for the board)

At this time, QCA4019, QCA9888, and QCA9984 are supported.
Extension to other chips should be straightforward.

The existing files, board-*.bin, are "grandfathered" as QCA4019.

The package name has been retained for compatability reasons.
At this time it DEPENDS:=@TARGET_ipq40xx, limiting its visibility.

Build-tested-on: asus_map-ac2200, alfa-network_ap120c-ac,
    avm_fritzbox-7530, avm_fritzrepeater-3000, engenius_eap1300,
    engenius_ens620ext, linksys_ea6350v3, qxwlan-e2600ac-c1/-c2

Signed-off-by: Jeff Kletsky <git-commits@allycomm.com>

jeffsf commented:
First, thank you all for the time and careful review. I appreciate the comments and the urging to unify and simplify where appropriate. 

Thanks for the pointer to `<<-EOF` -- learned a new trick!

I believe I've addressed all the comments and suggestions with the revised PR. 

# Commit: firmware/ipq-wifi: Extend for multi-chip boards

Most is covered in the commit message itself. Some of the thought behind implementation decisions follow. 

`make` syntax seemed more maintainable than trying to write a shell for loop and case statement as a single line in the recipe between the need for exact `;` placement and needing eight leading dollar signs. Handles multiple chips per board. Should be easily extensible to additional chips.

Existing `.bin` files "grandfathered" as QCA4019. 

In the two cases that the `.bin` files were named inconsistently, I chose to symlink them (rather than rename them) as it would preserve the git history of the files, not suggest that their contents had been changed, and that these should be comparatively short-lived files. Simple enough to rename if that is preferred.

Build test performed on all variants with all boards using those variants confirming the expected MD5 for the `board-2.bin` file(s) in the completed image, as well as the package.

# Commit: mtd: base-files: Unify dual-firmware devices (Linksys)

As with the previous commit, most is covered in the commit message itself or the commentary around the topic earlier in this PR.

All variants now use `/etc/init.d/bootcount` which is set for `START=99`. This is after `/etc/init.d/done` at `START=95`. 

All Linksys upgrade variants now try to handle the absence of `boot_part`. This can occur if the U-Boot environment has been erroneously modified or reset to default, such as on a CRC failure on the environment.


# Commit: ipq40xx: Add support for Linksys EA8300 (Dallas)

## General Comments

I've done further investigation and `firstboot` does appear to have a bug, one that impacts multiple, if not most, NAND-based devices. [FS#468](https://bugs.openwrt.org/index.php?do=details&task_id=468) has been updated with my findings that

```
sz = write(fd, &deadc0de, sizeof(deadc0de));
```

returns -1, indicating a write failure. I am, at this point, guessing that it is due to NAND flash, based on the reports all indicating UBIFS. 

I've reexamined the timing of the extraction of ART data and it was my mistake, it does now appear to be extracted in time for the "user helper" to load it on first boot. Thank you for speeding the extraction with that patch! My patch to examine the behavior is available at on the forum, [Showing (ath10k) Firmware Load Successes](https://forum.openwrt.org/t/showing-ath10k-firmware-load-successes/35550?u=jeff). Examining Linux 4.19 suggests that the fail-to-find warnings have been quelled which should make `dmessg` output more readable in the future.

My guess is that the CSR8811 is connected to the high-speed serial port that I have not been able to bring up yet. I've tried a couple of DTS variants without much success so far. Once connected, between the data sheet and notes from, for example, Laird, it appears that it can be recognized and managed by BlueZ and related tools. Finding the MAC in the OEM flash will hopefully be straightforward, or determining it from running OEM firmware, relative to the device's `hw_mac_addr`. OEM firmware shows `Bluetooth: Core ver 2.18` at boot. OEM boot log can be seen on the [OpenWrt wiki](https://openwrt.org/inbox/linksys/linksys_ea8300#oem_bootlog).


## Code-Specific Comments

Rather than flood the mailing list with a dozen or more "done" emails and not knowing the etiquite around "Resolve conversation", I've commented below where the recommendation was not implemented as described or clarification was requested, when not previously mentioned in the comments above. 

### package/kernel/mac80211/files/lib/wifi/mac80211.sh

No changes from `master` are being made to this file in revised version of this PR.

### package/firmware/ipq-wifi/Makefile:21

If `linksys_ea8300` is deemed suitable by the project (as opposed to `linksys_dallas` or the like), then I will submit these board files on confirmation.


### target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:96,126,172,229

Change to `cat /sys/class/net/eth0/address` done.

The OEM firmware assigns four, consecutive MAC addresses based on the single MAC burned in. Strangely enough, it reuses the same MAC address for both Ethernet phys and assigns the remaining three to the wireless radios. Sharing the same MAC address on the two Ethernet phys seems troublesome, for example in a situation where a switch is used to connect the modem to the router, as well as fanning out the LAN to various other devices. While it is unlikely that there would be a MAC collision on wireless with another device if a fifth Linksys MAC was "stolen", chosing the next in sequence and modifying it to be LA seemed a more appropriate choice.


### target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:100

`clk_ignore_unused` removed.

Yes, `ro` as I had to override the rootfstype and early `init` would not mount squashfs without `ro` present to override the `rw` in the OEM U-Boot command line. This permits flashing through the OEM GUI and simplifies "return to OEM".


### target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4019-linksys_ea8300.dts:136

Removed `0:` prefix on partitions. Lowercased all partition names with the exception of `ART`. `ART` is consistent with the other devices on the platform.

```
fgrep -i art files-4.14/arch/arm/boot/dts/qcom-ipq40* | fgrep 'label'
```

Easy to change if desired. 

### target/linux/ipq40xx/image/Makefile:237,238

Thanks, using `k` suffix as I couldn't get both `dd` and `bash` to agree on `m` vs. `M`.

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -115,8 +114,13 @@ case "$FIRMWARE" in
+		ath10kcal_patch_mac_crc $(macaddr_setbit_la $(macaddr_add "$(cat /sys/class/net/eth0/address)" 4))
+		;;
 	openmesh,a62)
-		ath10kcal_extract "0:ART" 36864 12064
```
Please don't change/break other boards (since this would also require a DTS update in qcom-ipq4019-a62.dts). But this is outside the scope of this patch.

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -150,12 +154,16 @@ case "$FIRMWARE" in
 	openmesh,a62 |\
 	qxwlan,e2600ac-c1 |\
 	qxwlan,e2600ac-c2)
-		ath10kcal_extract "0:ART" 4096 12064
```
Please don't change/break other boards (since this would also require a DTS update in qcom-ipq4019-a62.dts). But this is outside the scope of this patch.

chunkeey commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -203,12 +211,16 @@ case "$FIRMWARE" in
 	openmesh,a62 |\
 	qxwlan,e2600ac-c1 |\
 	qxwlan,e2600ac-c2)
-		ath10kcal_extract "0:ART" 20480 12064
```
Please don't change/break other boards (since this would also require a DTS update in qcom-ipq4019-a62.dts). But this is outside the scope of this patch.

chunkeey commented:
Have you sent out the boardfiles to ath10k yet?

jeffsf commented:
* Overzealous substitutions of 0:ART to ART resolved
* Rebased on `master` due conflicts from ee96fa15b1, _mvebu: use device-tree board detection_
* Board-name changes from ee96fa15b1 incorporated in `target/linux/mvebu/base-files/etc/init.d/bootcount`
* Clean build (including removal of `./tmp/`) re-checked on EA8300

I've been waiting on confirmation that "linksys,ea8300" was acceptable to the project (as opposed to "linksys,dallas" or the like). I take your comment about sending the board files upstream as confirmation and will do so based on that.

chunkeey commented:
> I've been waiting on confirmation that "linksys,ea8300" was acceptable to the project (as opposed to "linksys,dallas" or the like). I take your comment about sending the board files upstream as confirmation and will do so based on that.

I went back and forth whenever to name a device according to the customer product name or internal project name. I can't say what's the better version, from my experience I like the product name (EA8300) much more since it's less confusing when you are not that familiar with the device. 
Case in point would be in "mvebu". I can't tell you of the top of my head what dts the WRT1900AC v2 or WRT1900ACS has, even tough I've come across the terms on the forums several times. But because I don't have either I forget which is the Cobra or Shelby. On the other hand, for the people which are familiar they do like these project names because WRT1900ACS and WRT1900AC v2 are almost too similar, so it's easy to see why they want a better identifier for these devices, since they are different enough. So, I guess I recommend "Take your pick", I prefer the EA8300, but Dallas would be fine as well.

jeffsf commented:
A broken serial wire lead me to find that I had made an error in refactoring the DTS. Changed in both `files-4.14` and `files-4.19` relative to the last push
```
 &tlmm {
        serial_0_pins: serial0-pinmux {
                pins = "gpio16", "gpio17";
-               function = "blsp_uart1";
+               function = "blsp_uart0";
                bias-disable;
        };
```

Board files submitted upstream

http://lists.infradead.org/pipermail/ath10k/2019-May/013403.html
http://lists.infradead.org/pipermail/ath10k/2019-May/013404.html

Are there any further changes suggested, recommended, or required at this time?

chunkeey commented:
> Are there any further changes suggested, recommended, or required at this time?

No all good. Thanks for sticking around for so long!

chunkeey commented:
Merged. Thanks.
